samza-core/src/main/java/org/apache/samza/zk/ZkControllerImpl.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.zk;
+
+import org.I0Itec.zkclient.IZkChildListener;
+import org.I0Itec.zkclient.IZkDataListener;
+import org.apache.samza.SamzaException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+
+public class ZkControllerImpl implements ZkController {
+  private static final Logger LOG = LoggerFactory.getLogger(ZkControllerImpl.class);
+
+  private String processorIdStr;
+  private final ZkUtils zkUtils;
+  private final ZkListener zkListener;
+  private final ZkLeaderElector leaderElector;
+  private final ScheduleAfterDebounceTime debounceTimer;
+
+  public ZkControllerImpl (String processorIdStr, ZkUtils zkUtils, ScheduleAfterDebounceTime debounceTimer, ZkListener zkListener) {
+    this.processorIdStr = processorIdStr;
+    this.zkUtils = zkUtils;
+    this.zkListener = zkListener;
+    this.leaderElector = new ZkLeaderElector(this.processorIdStr, this.zkUtils);
+    this.debounceTimer = debounceTimer;
+
+    init();
+  }
+
+  @Override
+  public void register() {
+
+    // TODO - make a loop here with some number of attempts.
+    // possibly split into two method - becomeLeader() and becomeParticipant()
+    boolean isLeader = leaderElector.tryBecomeLeader();
+    if(isLeader) {
+      listenToProcessorLiveness();
```

Comment:
I think it should be moved into onBecomeLeader() - both the subscription for new processors and jobModel generation.